File: src/leetcode/m5/LongestPalindromicSubstring.go
package m5

// LeetCode 5
func longestPalindrome(s string) string {
	res := ""
	for i := 0; i < len(s); i++ {
		str1 := longest(s, i, i)
		str2 := longest(s, i, i+1)
		if len(str1) > len(res) {
			res = str1
		}
		if len(str2) > len(res) {
			res = str2
		}
	}
	return res
}

func longest(s string, l, r int) string {
	for l >= 0 && r < len(s) && s[l] == s[r] {
		l--
		r++
	}
	return s[l+1 : r]
}

File: src/leetcode/m5/LongestPalindromicSubstring.java
/*
 * Copyright (c) 2023-2025 THGEEK.COM. All rights reserved.
 */

package leetcode.m5;

/**
 * LeetCode 5
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2023/12/20 11:19
 */
public class LongestPalindromicSubstring {
    public String longestPalindrome(String s) {
        String res = "";
        for (int i = 0; i < s.length(); i++) {
            String res1 = longestPalindrome(s, i, i);
            String res2 = longestPalindrome(s, i, i + 1);
            String _res = res1.length() > res2.length() ? res1 : res2;
            res = _res.length() > res.length() ? _res : res;
        }
        return res;
    }

    private String longestPalindrome(String s, int l, int r) {
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        return s.substring(l + 1, r);
    }
}